core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -30,23 +35,98 @@
+    return wrap(catalog, true, expirationIntervalMillis);
+  }
+
+  public static Catalog wrap(Catalog catalog, Duration expirationInterval) {
+    return wrap(catalog, true, expirationInterval.toMillis());
   }
 
-  public static Catalog wrap(Catalog catalog, boolean caseSensitive) {
-    return new CachingCatalog(catalog, caseSensitive);
+  public static Catalog wrap(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis) {
+    return new CachingCatalog(catalog, caseSensitive, expirationIntervalMillis);
   }
 
-  private final Cache<TableIdentifier, Table> tableCache = Caffeine.newBuilder().softValues().build();
   private final Catalog catalog;
   private final boolean caseSensitive;
+  private final boolean expirationEnabled;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final long expirationIntervalMillis;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final Cache<TableIdentifier, Table> tableCache;
+
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis) {
+    this(catalog, caseSensitive, expirationIntervalMillis, Ticker.systemTicker());
+  }
 
-  private CachingCatalog(Catalog catalog, boolean caseSensitive) {
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected CachingCatalog(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis, Ticker ticker) {
     this.catalog = catalog;
     this.caseSensitive = caseSensitive;
+    // Set negative values to zero to avoid creating a Duration with a negative value
+    this.expirationIntervalMillis = expirationIntervalMillis <= 0 ? 0 : expirationIntervalMillis;
+    this.expirationEnabled = expirationIntervalMillis > 0;
+    this.tableCache = createTableCache(ticker);
+  }
+
+  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
+    Caffeine<TableIdentifier, Table> cacheBuilder = Caffeine
+        .newBuilder()
+        .softValues()
+        .removalListener(identLoggingRemovalListener);
+
+    // Only setup ticker and background cleanup tasks from expiration if table expiration is enabled.
+    if (expirationEnabled) {
+      return cacheBuilder
+          .writer(new CacheWriter<TableIdentifier, Table>() {
+            @Override
+            public void write(TableIdentifier tableIdentifier, Table table) {
+            }
+
+            @Override
+            public void delete(TableIdentifier tableIdentifier, Table table, RemovalCause cause) {
+              // On expiration (e.g. eviction due to `expireAfterAccess` or `expireAfterWrite` policies),
+              // perform any necessary cache cleanup so that subsequent catalog loads won't
+              // return stale metadata tables w.r.t. the underlying data tables they would return.
+              //
+              // TODO: Consider moving all table expiration here, including explicit calls to `invalidate`.
+              //       This call is asynchornous in a background thread though.
```

Comment:
I think you can remove the TODO. You're right to think about whether we should do that if it is async, and I don't think it is a good idea. If you invalidate an entry, then I like removing all metadata tables synchronously.

Reply:
Oh my bad I started updating this in another branch but then got sick and didn't open that yet.

I can move the changes from that branch here I guess.